### PR TITLE
Enable float8 CI on sm89

### PR DESCRIPTION
**Pull request by jainapurva**
The current CI/CD pipeline skipped float8 tests as they weren't compatible with the A10G GPUs. This PR adds a new CI job which runs on NVIDIA L4 Tensor Core GPUs for float8 tests. 

Fixes Issue: #575 

Test Plan : Compare the logs of Regression Test and Float8 Test to check for the skipped tests in Regression Test which are now being run on the new CI job of Float8 Test.

Sample of some test cases which were being skipped in CUDA Nightly in Regression Tests are running in Float8.

> - test/float8/test_numerics_integration.py::TestFloat8NumericsIntegrationTest::test_encoder_fw_bw
> - test/float8/test_inference_flows.py::TestHPTrainToFP8LinearInference::test_dynamic_fp8_mlp
> - test/float8/test_inference_flows.py::TestHPTrainToFP8LinearInference::test_static_fp8_mlp

Screenshot of logs of CUDA Nightly
image

Screenshot of logs of Float8 test
image

For full list of passed tests: https://gist.github.com/jainapurva/a5ddd17c809219151485de8d1708d078

### CODE REVIEW

**Comment by msaroufim** [comment]
So question here
1. Are these g6 machines cheaper than g5? Cause it sounds like count goes from 12 to 4
2. Also if the plan is launch an extra job for fp8 then the new job should only run fp8 tests, itll make your tests run a lot faster

File: .github/workflows/custom_test.yml [hunk 1]
```diff
@@ -0,0 +1,33 @@
+name: Custom Test
```
Comment (msaroufim) [inline .github/workflows/custom_test.yml:1]: yup this is cleaner

**Comment by msaroufim** [comment]
I noticed the fp8 test is not being triggered here so instead what I did was 
1. Copy pasted your updated regression test file
2. Deleted all strategies except the one that needs to run the g6 machines
3. Updated the test to only run fp8

So you can go ahead and make the updates here from this PR https://github.com/pytorch/ao/pull/603 I won't land 603

For Github Actions because you can't iterate fast because you need to wait on CI, would recommend copy pasting as much as possible or running tiny experiments where you in your branch temporarily delete all jobs except the one you're interested in iterating on. It makes iterating far more pleasant


Screenshot 2024-08-05 at 9 33 58 PM

File: .github/workflows/float8_test.yml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+on:
+  push:
+    branches:
+      - main
+  pull_request:
+    branches:
+      - main
+
+concurrency:
+  group: regression_test-${{ github.workflow }}-${{ github.ref == 'refs/heads/main' && github.run_number || github.ref }}
```
Comment (msaroufim) [inline .github/workflows/float8_test.yml:12]: typo on my end,  rename this to fp8

**Review by msaroufim — DISMISSED** [review]

File: test/float8/test_base.py [hunk 3]
```diff
@@ -55,7 +55,7 @@
 torch.manual_seed(0)
 
 is_H100 = torch.cuda.is_available() and torch.cuda.get_device_capability() >= (9, 0)
-
+is_cuda_8_9 = torch.cuda.is_available() and torch.cuda.get_device_capability() >= (8, 9)
```
Comment (msaroufim) [inline test/float8/test_base.py:58]: should we this anywhere? It's not super clear to me which additional tests are now running relative to the original regression test job

Comment (vkuzo) [reply test/float8/test_base.py:58]: right, the next step would be:
1. change the tests which are gating on `IS_H100` to gate on `is_cuda_8_9` instead, in this file and all other float8 test files.  Exception: any tests which depend on rowwise gemms need to keep `IS_H100`.
2. inspect the CI logs and verify that the tests requiring this capability are no longer skippe.

Comment (jainapurva) [reply test/float8/test_base.py:58]: Ok, working on it

Comment (vkuzo) [reply test/float8/test_base.py:58]: can we delete `is_H100` if it's not used, for all of these files?

File: test/float8/test_numerics_integration.py [hunk 4]
```diff
@@ -89,7 +89,7 @@ class TestFloat8NumericsIntegrationTest:
         "scaling_type_grad_output",
         [ScalingType.DELAYED, ScalingType.DYNAMIC],
     )
-    @pytest.mark.skipif(not is_H100, reason="requires H100 GPU")
+    @pytest.mark.skipif(not is_cuda_8_9, reason="requires H100 GPU")
```
Comment (vkuzo) [inline test/float8/test_numerics_integration.py:92]: nit: update text description as well?

File: .github/workflows/float8_test.yml [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+jobs:
+  test:
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          - name: SM-89
+            runs-on: amz2023.linux.g6.4xlarge.experimental.nvidia.gpu
+            torch-spec: '--pre torch==2.5.0.dev20240728+cu121 --index-url https://download.pytorch.org/whl/nightly/cu121'
```
Comment (msaroufim) [inline .github/workflows/float8_test.yml:26]: why is this pinned?

File: test/float8/test_base.py (hunk 3 above)
Comment (jainapurva) [reply test/float8/test_base.py:58]: Removed for unused files

File: .github/workflows/custom_test.yml (hunk 1 above)
Comment (jainapurva) [reply .github/workflows/custom_test.yml]: Done

**Review by msaroufim — APPROVED** [review]

**Pull request merged**